include/swift/AST/Evaluator.h

```diff
@@ -0,0 +1,264 @@
+//===--- Evaluator.h - Request Evaluator ------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the Evaluator class that evaluates and caches
+// requests.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_AST_EVALUATOR_H
+#define SWIFT_AST_EVALUATOR_H
+
+#include "swift/AST/AnyRequest.h"
+#include "swift/Basic/AnyValue.h"
+#include "swift/Basic/Defer.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/Optional.h"
+#include <tuple>
+#include <type_traits>
+#include <vector>
+
+namespace swift {
+
+using llvm::Optional;
+using llvm::None;
+
+class DiagnosticEngine;
+
+/// Evaluation engine that evaluates and caches "requests", checking for cyclic
+/// dependencies along the way.
+///
+/// Each request is a function object that accepts a reference to the evaluator
+/// itself (through which it can request other values) and produces a
+/// value. That value can then be cached by the evaluator for subsequent access,
+/// using a policy dictated by the request itself.
+///
+/// The evaluator keeps track of all in-flight requests so that it can detect
+/// and diagnose cyclic dependencies.
+///
+/// Each request should be its own function object, supporting the following
+/// API:
+///
+///   - Copy constructor
+///   - Equality operator (==)
+///   - Hashing support (hash_value)
+///   - TypeID support (see swift/Basic/TypeID.h)
+///   - The output type (described via a nested type OutputType), which
+///     must itself by a value type that supports TypeID.
+///   - Evaluation via the function call operator:
+///       OutputType operator()(Evaluator &evaluator) const;
+///   - Cycle breaking and diagnostics operations:
+///
+///       void diagnoseCycle(DiagnosticEngine &diags) const;
+///       void noteCycleStep(DiagnosticEngine &diags) const;
+///       OutputType breakCycle() Const;
+///   - Caching policy:
+///
+///     static const bool isEverCached;
+///
+///       When false, the request's result will never be cached. Note that this
+///       also precludes cycle detection.  When true, the result will be
```

Comment:
Why does not-caching preclude cycle detection?

Reply:
Ah, it doesn't have to. I'm using the cache to detect the presence of a cycle, but there's no reason I have to do that... I have the stack of results to be satisfied and can unique there. Thanks!